### PR TITLE
feat: support hide the authentication header in basic-auth with  a config

### LINKED ISSUE #5900: request help: support hide the authentication header in basic-auth and key-auth plugins
### Issue description

Hello,
When using base-auth plugins, I don't want to pass the `Authentication` header to upstream.
IMO, there is no need for upstream to perceive these authentication headers.

Now, I use `proxy-rewrite` plugin to rewrite the `Authentication` head like this:

```
 "plugins": {
        "basic-auth": {},
        "proxy-rewrite": {
            "headers": {
                "Authorization": ""
            }
        }
    },
```

I think we can add a config in the `basic-auth` plugin to hide the auth head.
That will be more convenient in using these plugins. So do `key-auth` plugin.

### Environment

- apisix version (cmd: `apisix version`): 2.11.0
- OS (cmd: `uname -a`): 
- OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`):
- etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API):
- apisix-dashboard version, if have:
- the plugin runner version, if the issue is about a plugin runner (cmd: depended on the kind of runner):
- luarocks version, if the issue is about installation (cmd: `luarocks --version`):

### PR DESCRIPTION
**Pull request by mangoGoForward**
### What this PR does / why we need it:
Resolves #5900

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/plugins/basic-auth.lua [hunk 1]
```diff
@@ -161,6 +170,11 @@ function _M.rewrite(conf, ctx)
         return 401, { message = "Password is error" }
     end
 
+    -- 5. hide `Authentication` header if `hide_auth_header` is `true`
+    if conf.hide_auth_header == true then
+        core.response.set_header("Authentication", "")
```
Comment (spacewander) [inline apisix/plugins/basic-auth.lua:175]: It seems you misunderstand the original issue. We want to hide the request header.

Comment (mangoGoForward) [reply apisix/plugins/basic-auth.lua]: My fault.  Done.

File: docs/zh/latest/plugins/basic-auth.md [hunk 2]
```diff
@@ -43,6 +43,7 @@ title: basic-auth
 | -------- | ------ | ------ | ------ | ------ | ------------------------------------------------------------------------------------------------------------------ |
 | username | string | 必须   |        |        | 不同的 `consumer` 对象应有不同的值，它应当是唯一的。不同 consumer 使用了相同的 `username` ，将会出现请求匹配异常。 |
 | password | string | 必须   |        |        | 用户的密码                                                                                                         |
+| hide_auth_header | boolean | 可选    | true   |       | 是否将 Authentication 请求头返回给客户端.                                                                                             |
```
Comment (leslie-tsang) [inline docs/zh/latest/plugins/basic-auth.md:46]:
```suggestion
| hide_auth_header | boolean | 可选    | false   |       | 是否将 Authentication 请求头传递给 upstream。                                                                                            |
```

File: docs/en/latest/plugins/basic-auth.md [hunk 3]
```diff
@@ -39,10 +39,11 @@ For more information on Basic authentication, refer to [Wiki](https://en.wikiped
...
+| Name             | Type    | Requirement | Default | Valid | Description                                                                                                                                                      |
+| --------         | ------  | ----------- | ------- | ----- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| username         | string  | required    |         |       | Different `consumer` should have different value which is unique. When different `consumer` use a same `username`, a request matching exception would be raised. |
+| password         | string  | required    |         |       | the user's password                                                                                                                                              |
+| hide_auth_header | boolean | optional    | true    |       | Whether to return the Authentication request headers to the client.                                                                                             |
```
Comment (leslie-tsang) [inline docs/en/latest/plugins/basic-auth.md:46]:
```suggestion
| Name             | Type    | Requirement | Default | Valid | Description                                                                                                                                                      |
| --------         | ------  | ----------- | ------- | ----- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| username         | string  | required    |         |       | Different `consumer` should have different value which is unique. When different `consumer` use a same `username`, a request matching exception would be raised. |
| password         | string  | required    |         |       | the user's password                                                                                                                                              |
| hide_auth_header | boolean | optional    | true    |       | Whether to return the Authentication request headers to the upstream.                                                                                             |
```

File: apisix/plugins/basic-auth.lua [hunk 4]
```diff
@@ -39,6 +44,10 @@ local consumer_schema = {
     properties = {
         username = { type = "string" },
         password = { type = "string" },
+        hide_auth_header = {
+            type = "boolean",
+            default = true,
```
Comment (leslie-tsang) [inline apisix/plugins/basic-auth.lua:49]:
```suggestion
            default = false,
```
Need to discuss it in maillist if use `true` as default value

File: apisix/plugins/basic-auth.lua [hunk 5]
```diff
@@ -30,7 +30,12 @@ local consumers_lrucache = core.lrucache.new({
 local schema = {
     type = "object",
     title = "work with route or service object",
-    properties = {},
+    properties = {
+        hide_auth_header = {
+            type = "boolean",
+            default = true,
```
Comment (tokers) [inline apisix/plugins/basic-auth.lua:36]: I think the default value should be false.

File: apisix/plugins/basic-auth.lua [hunk 6]
```diff
@@ -161,6 +170,11 @@ function _M.rewrite(conf, ctx)
         return 401, { message = "Password is error" }
     end
 
+    -- 5. hide `Authentication` request header if `hide_auth_header` is `true`
+    if conf.hide_auth_header == true then
+        core.request.set_header(ctx, "Authentication", "")
+    end
+
```
Comment (leslie-tsang) [inline apisix/plugins/basic-auth.lua:177]: The original issue seems to avoid to send the header to upstream.

**Review by leslie-tsang — CHANGES_REQUESTED** [review]

File: apisix/plugins/basic-auth.lua (hunk 5 above)
Comment (leslie-tsang) [inline apisix/plugins/basic-auth.lua:36]:
```suggestion
            default = true,
```
Ditto

Comment (mangoGoForward) [reply apisix/plugins/basic-auth.lua]: Done.

File: docs/zh/latest/plugins/basic-auth.md (hunk 2 above)
Comment (mangoGoForward) [reply docs/zh/latest/plugins/basic-auth.md]: Done.

File: docs/en/latest/plugins/basic-auth.md (hunk 3 above)
Comment (mangoGoForward) [reply docs/en/latest/plugins/basic-auth.md]: Done.

File: apisix/plugins/basic-auth.lua (hunk 6 above)
Comment (mangoGoForward) [reply apisix/plugins/basic-auth.lua]: Yes, use `core.request.set_header(ctx, "Authentication", nil)` can avoid.

File: apisix/plugins/basic-auth.lua (hunk 5 above)
Comment (mangoGoForward) [reply apisix/plugins/basic-auth.lua]: Done.

File: apisix/plugins/basic-auth.lua [hunk 7]
```diff
@@ -30,7 +30,12 @@ local consumers_lrucache = core.lrucache.new({
 local schema = {
     type = "object",
     title = "work with route or service object",
-    properties = {},
+    properties = {
+        hide_auth_header = {
```
Comment (spacewander) [inline apisix/plugins/basic-auth.lua:34]: `hide_credentials` would be better? Kong uses this field in their basic-auth.

File: apisix/plugins/basic-auth.lua [hunk 8]
```diff
@@ -39,6 +44,10 @@ local consumer_schema = {
     properties = {
         username = { type = "string" },
         password = { type = "string" },
+        hide_auth_header = {
```
Comment (spacewander) [inline apisix/plugins/basic-auth.lua:47]: We don't need to configure it in the consumer

File: docs/en/latest/plugins/basic-auth.md [hunk 9]
```diff
@@ -39,10 +39,11 @@ For more information on Basic authentication, refer to [Wiki](https://en.wikiped
...
+| Name             | Type    | Requirement | Default | Valid | Description                                                                                                                                                      |
+| --------         | ------  | ----------- | ------- | ----- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| username         | string  | required    |         |       | Different `consumer` should have different value which is unique. When different `consumer` use a same `username`, a request matching exception would be raised. |
+| password         | string  | required    |         |       | the user's password                                                                                                                                              |
+| hide_auth_header | boolean | optional    | false   |       | Whether to return the Authentication request headers to the upstream.                                                                                             |
```
Comment (spacewander) [inline docs/en/latest/plugins/basic-auth.md:46]: Let's distinguish route conf from the consumer's like this one: https://github.com/apache/apisix/blob/master/docs/en/latest/plugins/ldap-auth.md#attributes

File: apisix/plugins/basic-auth.lua (hunk 7 above)
Comment (mangoGoForward) [reply apisix/plugins/basic-auth.lua]: Done.

File: apisix/plugins/basic-auth.lua (hunk 8 above)
Comment (mangoGoForward) [reply apisix/plugins/basic-auth.lua]: Done.

File: docs/en/latest/plugins/basic-auth.md (hunk 9 above)
Comment (mangoGoForward) [reply docs/en/latest/plugins/basic-auth.md]: Done.

File: apisix/plugins/basic-auth.lua [hunk 10]
```diff
@@ -161,6 +166,11 @@ function _M.rewrite(conf, ctx)
         return 401, { message = "Password is error" }
     end
 
+    -- 5. hide `Authentication` request header if `hide_credentials` is `true`
+    if conf.hide_credentials == true then
```
Comment (tzssangglass) [inline apisix/plugins/basic-auth.lua:170]:
```suggestion
    if conf.hide_credentials then
```

is ok?

**Comment by tzssangglass** [comment]
and need to add test cases to cover this. I see that you have test cases in your previous submission records……

File: docs/en/latest/plugins/basic-auth.md [hunk 11]
```diff
@@ -39,11 +39,18 @@ For more information on Basic authentication, refer to [Wiki](https://en.wikiped
...
 | password         | string  | required    |         |       | the user's password                                                                                                                                              |
-| hide_auth_header | boolean | optional    | false   |       | Whether to return the Authentication request headers to the upstream.                                                                                             |
+
+For route side:
+
+| Name             | Type    | Requirement | Default | Valid | Description                                                                                                                                                      |
+| --------         | ------  | ----------- | ------- | ----- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| hide_credentials | boolean | optional    | false   |       | Whether to return the Authentication request headers to the upstream.                                                                                            |
```
Comment (spacewander) [inline docs/en/latest/plugins/basic-auth.md:53]:
```suggestion
| hide_credentials | boolean | optional    | false   |       | Whether to pass the Authentication request headers to the upstream.                                                                                            |
```

File: apisix/plugins/basic-auth.lua (hunk 10 above)
Comment (mangoGoForward) [reply apisix/plugins/basic-auth.lua]: Done.

File: docs/en/latest/plugins/basic-auth.md (hunk 11 above)
Comment (mangoGoForward) [reply docs/en/latest/plugins/basic-auth.md]: Done.

**Comment by mangoGoForward** [comment]
> and need to add test cases to cover this. I see that you have test cases in your previous submission records……

Because I misunderstand the original issue, and those test cases seems incorrect, so I removed. Now test cases updated in 210c9d5, please give me some suggestion if you have a free times, Thanks.

File: t/plugin/basic-auth.t [hunk 12]
```diff
@@ -356,3 +356,55 @@ GET /t
...
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST 16: verify
```
Comment (tokers) [inline t/plugin/basic-auth.t:400]: Can we strengthen this case that we can make sure the header is really removed, rather than just checking out the log.

Comment (mangoGoForward) [reply t/plugin/basic-auth.t:400]: I am ashamed not found such parse in project of test-nginx, how could we detect request header?

Comment (leslie-tsang) [reply t/plugin/basic-auth.t:400]: Try `response_headers` ?

Comment (mangoGoForward) [reply t/plugin/basic-auth.t:400]: The `response_headers` data section can be used to validate response header entries, but `Authorization` is request header, so I don't know whether `response_headers` can used in this test case.

Comment (tzssangglass) [reply t/plugin/basic-auth.t:400]: we can mock an upstream, and check headers in upstream, take a look at: https://github.com/apache/apisix/blob/70174d58cd4248f55dada5c2bc7396cdf73972c7/t/plugin/proxy-mirror.t#L35-L55

Comment (mangoGoForward) [reply t/plugin/basic-auth.t:440]: Done. please see cb53193

File: t/plugin/basic-auth.t [hunk 13]
```diff
@@ -20,6 +20,46 @@ repeat_each(2);
 no_long_string();
 no_root_location();
 no_shuffle();
+log_level('info');
+worker_connections(1024);
+
+add_block_preprocessor(sub {
+    my ($block) = @_;
+
+    my $http_config = $block->http_config // <<_EOC_;
```
Comment (spacewander) [inline t/plugin/basic-auth.t:29]: We already have an upstream for this, see: https://github.com/apache/apisix/blob/70174d58cd4248f55dada5c2bc7396cdf73972c7/t/lib/server.lua#L392

Use /echo is enough.

**Comment by imjoey** [comment]
@mangoGoForward Hi, as @spacewander suggested, you could use `/echo` instead of `httpbin.org` to check the headers. Please refer to https://github.com/apache/apisix/blob/d4a7ea208fa748ac52a9a77d82d018e1fb971307/t/plugin/zipkin.t#L419.

File: t/plugin/basic-auth.t (hunk 13 above)
Comment (mangoGoForward) [reply t/plugin/basic-auth.t]: Done.

**Comment by mangoGoForward** [comment]
> @mangoGoForward Hi, as @spacewander suggested, you could use `/echo` instead of `httpbin.org` to check the headers. Please refer to
> 
> https://github.com/apache/apisix/blob/d4a7ea208fa748ac52a9a77d82d018e1fb971307/t/plugin/zipkin.t#L419
> 
> .

Thanks, have changed, please see 25602a7.

File: docs/en/latest/plugins/basic-auth.md [hunk 14]
```diff
@@ -39,10 +39,18 @@ For more information on Basic authentication, refer to [Wiki](https://en.wikiped
...
-| password | string | required    |         |       | the user's password                                                                                                                                              |
+For consumer side:
+
+| Name             | Type    | Requirement | Default | Valid | Description                                                                                                                                                      |
+| --------         | ------  | ----------- | ------- | ----- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| username         | string  | required    |         |       | Different `consumer` should have different value which is unique. When different `consumer` use a same `username`, a request matching exception would be raised. |
+| password         | string  | required    |         |       | the user's password                                                                                                                                              |
```
Comment (imjoey) [inline docs/en/latest/plugins/basic-auth.md:47]: @mangoGoForward  IMHO there's no need to change this. How about revert it?

**Review by imjoey — CHANGES_REQUESTED** [review]
@mangoGoForward Please revert some changes in English docs, others look good to me. Thanks.

File: docs/en/latest/plugins/basic-auth.md (hunk 14 above)
Comment (mangoGoForward) [reply docs/en/latest/plugins/basic-auth.md]: Done.

File: t/plugin/basic-auth.t [hunk 15]
```diff
@@ -20,6 +20,7 @@ repeat_each(2);
 no_long_string();
 no_root_location();
 no_shuffle();
+log_level('info');
```
Comment (imjoey) [inline t/plugin/basic-auth.t:23]: @mangoGoForward please take a check whether this change is needed. Thanks.

Comment (mangoGoForward) [reply t/plugin/basic-auth.t]: Seems unnecessary, removed.

**Comment by spacewander** [comment]
@mangoGoForward 
Please make lint pass.

**Comment by mangoGoForward** [comment]
> @mangoGoForward Please make lint pass.

That's strange, it's good when  I run `reindex` on local(MacOS) and Centos environment, please see the screenshot below:

image

I will test again on ubantu later, If you know where the error location, please help me to improve, Thanks.

**Comment by mangoGoForward** [comment]
I run `reindex` to check test code style on `Ubuntu Server 20.04 LTS`, also is right, so strangely. please see:

image

**Comment by tzssangglass** [comment]
Strange, I had no problem checking the format of this test case.

**Review by spacewander — APPROVED** [review]

**Comment by mangoGoForward** [comment]
> Strange, I had no problem checking the format of this test case.

I have run workflow in my fork repo, it also failed.

image

**Comment by imjoey** [comment]
@mangoGoForward  The changes look good to me. Could you explain why you committed several other commits? Thanks.

**Comment by mangoGoForward** [comment]
> @mangoGoForward The changes look good to me. Could you explain why you committed several other commits? Thanks.

I want to trigger the actions of my forked repo, so I commit a PR to my repo's master branch, and got file conflict when merged from master, so i commit 8440045 to resolved. but the master branch's code  is polluted, so I revert the commit to 49b7850, and I also checked those file changes, it is same as 49b7850, I'm sorry about troubled your review work.

**Comment by imjoey** [comment]
@spacewander @mangoGoForward  The CI job failure of `linux_openresty on ubuntu-18.04` seemed not related to the code and I restarted that job. I would prefer to leave an approval in prior and merge this after that job passed. Thanks.

**Review by imjoey — APPROVED** [review]
The changes look good to me.

**Comment by lijing-21** [comment]
Hi @mangoGoForward , thank you for your contribution!

Here is the Contributor T-shirt form[1], if you're interested, kindly take a look :)

[1] https://github.com/apache/apisix/blob/master/CONTRIBUTING.md#contributor-t-shirt

**Comment by mangoGoForward** [comment]
> 

Thanks.

**Comment by mangoGoForward** [comment]
> Hi @mangoGoForward , thank you for your contribution!
> 
> Here is the Contributor T-shirt form[1], if you're interested, kindly take a look :)
> 
> [1] https://github.com/apache/apisix/blob/master/CONTRIBUTING.md#contributor-t-shirt

Thanks.

**Pull request merged**